DecipherNow/helm-charts
# [release-2.2] Refactor secrets to use cert/secret files, enable global certs, multiple docker registries

@**Please provide enough information so that others can review your pull request:**

1. Explain the **details** for making this change. What existing problem does the pull request solve? If it resolves an existing issue, be sure to use a Github [keyword](https://help.github.com/en/articles/closing-issues-using-keywords) to automatically close it.

Addresses issue 591 (https://github.com/DecipherNow/helm-charts/issues/591).  This pr enables certificates and some secrets to be passed in as files or from the values file.  Also allows a global certificate to be deployed which replaces many of the independently configurable certificate.

2. What **changes to values.yaml** are required?

**in secrets/values.yaml:**
- i added a from_file section to many certs and secrets parent keys
```
from_file:
  enabled: true
  path: files/certs/ingress
```
- jwt (internal and regular) I changed the certificate keys from `jwt.key.pem` -> `jwt_key_pem`, `jwt.trust.pem` -> `jwt_key_pem`, and `server.cert.pem` -> `server_cert_pem`

- changed ingress and egress values from a list into separate entries

**credentials.yaml** (generated)
- i made the docker login info a list so you can add multiple repo login info

```yaml
dockerCredentials:
  - registry: docker.repo1
    email: user@email.com
    username: user@email.com
    password: p@ssWord1
  - registry: docker.repo2
    email: user@email.com
    username: user@email.com
    password: p@ssWord2
```

**sense/values.yaml**
- i changed the slo secret name from `ca.pem` -> `ca.crt`

3. Have you documented any additional setup steps or configurations options?

Documentation has been added to [secrets/README.md](secrets/README.md)

4. Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.

This pr brings many changes that i have tested individually and in groups.
The basic procedure is to run:
 - if you would like to start fresh run `make destroy` and `make k3d`; otherwise run `make delete` and `make remove-secrets`

- `make credentials`
- `make secrets`
- `make install`

**Test adding multiple docker repo auths in the docker.secret**

*Test the template to build credentials works*
- run `make credentials if you do not already have a `credentials.yaml`
- Add an additional repo to credentials.yaml (look in the secrets readme for an example)
- run `make secrets`
- ensure the secret is created by running`kubectl get secret docker.secret -o yaml --export`
-- copy the b64 encoded part after `.dockerconfigjson: <the b64sring you wanna copy>` and decode that with `echo <the b64sring  you wanna copy> | base64 -D` check the print out to see if it makes sense. 

should be 
```
{"auths": {"docker.production.deciphernow.com": {"auth": "base64usernameandpassordstring"},"docker-dev.production.deciphernow.com": {"auth": "base64usernameandpassordstring"}}}
```
where "base64usernameandpassordstring"  can be decoded in a similar way for your creds.

*Test the created docker secret is valid* (assumes you have pull access to the docker-dev docker repo)
- switch a service's image in the values file of that service to the second repo we added
-- i used edge cause thats quick.  in `edge/values.yaml` change the version to 1.3.0 and the image from `docker.production.....`to `docker-dev.production.....`
-- `cd edge`
-- run `make edge`
-- ensure image is pulled `kc get pods`
-- clean up with `cd .. && make uninstall && make remove-secrets`

**Test Global Certs**

*Test values files certs works:*
- in `global.yaml` set `global.global_certs.enabled=true`
- `make secrets`
- `make install`
- If everything works you should see a working dashboard at https:localhost:30000
- clean up with `make uninstall` && `make remove-secrets`

*Test from_file works*
- like above in`global.yaml` set `global.global_certs.enabled=true` but also set `global.global_certs.from_file.enabled=true` and leave `global.global_certs.from_file.path=files/certs/global`
- `make secrets`
- `make install`
- If everything works you should see a working dashboard at https:localhost:30000
- clean up with `make uninstall && make remove-secrets`
- set `global.global_certs.enabled=false` in globals.yaml 

**Test other cert from_file**
Applicable for controlApi, control, dashboard, edge (ingress and egress), jwt (internal and external), mongo (internal and external), postgres, sidecar, and slo. - aka whats listed in `files/certs`
- copy certs from values file into their own files in the path defined at `from_file.path`
- set `from_file.enabled=true` in values.yaml for the service youre testing
- run `make secrets`
- run `make install`
- If everything works you should see a working dashboard at https:localhost:30000
- clean up with `make uninstall && make remove-secrets``

**Test secrets from_file**
Applicable for jwt (internal and external). - aka whats listed in `files/secrets`
- copy secrets (not the api key) from values file into their own files in the path defined at `from_file.path`
- set `from_file.enabled=true` in values.yaml for the service youre testing
- run `make secrets`
- run `make install`
- If everything works you should see a working dashboard at https:localhost:30000

Phew that was allot, give yourself a pat on the back.

- clean up with `make uninstall && make remove-secrets``
